mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala

```diff
@@ -61,20 +67,70 @@ class LogisticRegressionModel (
 
   override protected def predictPoint(dataMatrix: Vector, weightMatrix: Vector,
       intercept: Double) = {
-    val margin = weightMatrix.toBreeze.dot(dataMatrix.toBreeze) + intercept
-    val score = 1.0 / (1.0 + math.exp(-margin))
-    threshold match {
-      case Some(t) => if (score > t) 1.0 else 0.0
-      case None => score
+    // If dataMatrix and weightMatrix have the same dimension, it's binary logistic regression.
+    if (dataMatrix.size == weightMatrix.size) {
+      val margin = dot(weights, dataMatrix) + intercept
+      val score = 1.0 / (1.0 + math.exp(-margin))
+      threshold match {
+        case Some(t) => if (score > t) 1.0 else 0.0
+        case None => score
+      }
+    } else {
+      val dataWithBiasSize = weightMatrix.size / (nClasses - 1)
+      val dataWithBias = if(dataWithBiasSize == dataMatrix.size) {
+        dataMatrix
+      }  else {
+        assert(dataMatrix.size + 1 == dataWithBiasSize)
+        MLUtils.appendBias(dataMatrix)
+      }
+
+      val margins = Array.ofDim[Double](nClasses)
+
+      val weightsArray = weights match {
+      case dv: DenseVector => dv.values
+      case _ =>
+        throw new IllegalArgumentException(
+          s"weights only supports dense vector but got type ${weights.getClass}.")
+      }
+
+      var i = 0
+      while (i < nClasses - 1) {
```

Comment:
This may be minor but I feel like this could be more compact by creating `margins` by `map`ping `0 until nClasses` rather than unrolling it. The outer loop is not a tight loop here.